### ISSUE TITLE
Create a new type after deleting a property from object

### ISSUE THREAD

**Issue by bezreyhan** [microsoft/TypeScript#42069]
**TypeScript Version:**  v4.13

**Search Terms:** delete property from object new type

**Code**

```ts
interface ApiData {
    foo?: string | null;
}

interface Props {
    foo?: string;
}

function test(props: Props) {
    console.log(props.foo);
} 

const bar: ApiData = {
    foo: null
}

if (bar.foo === null) {
    delete p.foo;
}

test(bar);
```

**Expected behavior:**
No type error should occur when calling `test` with the argument `bar` since we have deleted `foo` from `bar` when `foo` is `null`.

**Actual behavior:**
The following error is reported by TS: `Argument of type 'ApiData' is not assignable to parameter of type 'Props'.
  Types of property 'foo' are incompatible.
    Type 'string | null | undefined' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.(2345)`

**Playground Link: https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIIAdgBE5jsgbwChlTkYB7CgfgC5kBnMKUAc2QB9kQBXAGz4BuIgF8iRUJFiIUABSgV0DQiTKUa9JixCthYojB4gEYYBRDJITABToFS+vMUMAlCrLIE5hhT4QAdHwUrLb2DP7qLnrI4l4gTMgARnBQ9BjYuPgAvO5qVPS8AqLiwDDI1slQEVTIWXXc-HxuxB4AJhB+kEkp1RR64lZgFSlRQA** 

**Related Issues:** https://github.com/microsoft/TypeScript/issues/13783
